resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/tpr/TPRCrudCallsImpl.scala

```diff
@@ -0,0 +1,320 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.kubernetes.tpr
+
+import java.util.concurrent.{ThreadPoolExecutor, TimeUnit}
+
+import io.fabric8.kubernetes.client.{BaseClient, KubernetesClient}
+import okhttp3.{HttpUrl, MediaType, OkHttpClient, Request, RequestBody, Response}
+import okio.{Buffer, BufferedSource}
+import org.json4s.{DefaultFormats, Formats}
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods.{compact, parse, pretty, render}
+import org.json4s.jackson.Serialization.{read, write}
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{blocking, Future, Promise}
+import scala.util.control.Breaks.{break, breakable}
+
+import org.apache.spark.deploy.kubernetes.constants._
+import org.apache.spark.internal.Logging
+import org.apache.spark.SparkException
+import org.apache.spark.util.ThreadUtils
+
+private[spark] trait TPRCrudCalls {
+  /**
+   * This trait contains currently acceptable operations on the SparkJob Resource
+   * CRUD + Watch
+   */
+  def createJobObject(name: String, keyValuePairs: Map[String, Any]): Unit
+  def deleteJobObject(tprObjectName: String): Unit
+  def getJobObject(name: String): SparkJobState
+  def updateJobObject(name: String, value: String, fieldPath: String): Unit
+  def watchJobObject(): Future[WatchObject]
+}
+
+private[spark] case class Metadata(name: String,
+    uid: Option[String] = None,
+    labels: Option[Map[String, String]] = None,
+    annotations: Option[Map[String, String]] = None)
+
+private[spark] case class SparkJobState(apiVersion: String,
+    kind: String,
+    metadata: Metadata,
+    status: Map[String, Any])
+
+private[spark] case class WatchObject(`type`: String, `object`: SparkJobState)
+
+/**
+ * Prereq - This assumes the kubernetes API has been extended using a TPR of name: Spark-job
+ * See conf/kubernetes-custom-resource.yaml for a model
+ *
+ * CRUD + Watch Operations on SparkJob Resource
+ *
+ * This class contains all CRUD+Watch implementations performed
+ * on the SparkJob Resource used to expose the state of a spark job
+ * in kubernetes (visible via kubectl or through the k8s dashboard).
+ */
+private[spark] class TPRCrudCallsImpl(k8sClient: KubernetesClient,
+    kubeToken: Option[String] = None) extends Logging with TPRCrudCalls {
+  private[spark] val kubeMaster: String = k8sClient.getMasterUrl().toString
+  private val httpClient: OkHttpClient =
+    extractHttpClientFromK8sClient(k8sClient.asInstanceOf[BaseClient])
+  private val namespace: String = k8sClient.getNamespace
+  private var watchSource: BufferedSource = _
+  private lazy val buffer = new Buffer()
```

Comment:
This is perhaps better to instantiate up front anyways - since it's an empty collection there shouldn't be much overhead.